using_restricted_array.rb

```diff
@@ -7,41 +7,88 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+    i = 0
+    until array[i] == nil
+      puts array[i]
+      i += 1
+    end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  until array[length - 1] < 0
```

Comment:
This should be `until length  <= 0` Because you want to run the loop length number of times. The number of times you run the loop should not depend on the value at an index aka `array[length - 1]`. If we had all negative integers in the array, the loop will not run even once.